org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/JDTLanguageServer.java

```diff
@@ -164,9 +164,70 @@ public void disconnectClient() {
+		}
+		if (preferenceManager.getClientPreferences().isDefinitionDynamicRegistered()) {
+			JavaLanguageServerPlugin.getInstance().registerCapability(Preferences.DEFINITION_ID, Preferences.DEFINITION);
+		}
+		if (preferenceManager.getClientPreferences().isHoverDynamicRegistered()) {
+			JavaLanguageServerPlugin.getInstance().registerCapability(Preferences.HOVER_ID, Preferences.HOVER);
+		}
+		if (preferenceManager.getClientPreferences().isReferencesDynamicRegistered()) {
+			JavaLanguageServerPlugin.getInstance().registerCapability(Preferences.REFERENCES_ID, Preferences.REFERENCES);
+		}
+		if (preferenceManager.getClientPreferences().isDocumentHighlightDynamicRegistered()) {
+			JavaLanguageServerPlugin.getInstance().registerCapability(Preferences.DOCUMENT_HIGHLIGHT_ID, Preferences.DOCUMENT_HIGHLIGHT);
+		}
 		if (preferenceManager.getClientPreferences().isWorkspaceFoldersSupported()) {
 			registerCapability(Preferences.WORKSPACE_CHANGE_FOLDERS_ID, Preferences.WORKSPACE_CHANGE_FOLDERS);
 		}
+		// we do not have the user setting initialized yet at this point but we should
+		// still call to enable defaults in case client does not support configuration changes
+		syncCapabilitiesToSettings();
+		try {
+			boolean autoBuildChanged = pm.setAutoBuilding(preferenceManager.getPreferences().isAutobuildEnabled());
+			buildWorkspace(autoBuildChanged);
+		} catch (CoreException e) {
+			JavaLanguageServerPlugin.logException(e.getMessage(), e);
+		}
+	}
+
+	/**
+	 * Toggles the server capabilities according to user preferences.
+	 */
+	private void syncCapabilitiesToSettings() {
+		if (preferenceManager.getClientPreferences().isFormattingDynamicRegistrationSupported()) {
+			toggleCapability(preferenceManager.getPreferences().isJavaFormatEnabled(), Preferences.FORMATTING_ID, Preferences.TEXT_DOCUMENT_FORMATTING, null);
+		}
+		if (preferenceManager.getClientPreferences().isRangeFormattingDynamicRegistrationSupported()) {
+			toggleCapability(preferenceManager.getPreferences().isJavaFormatEnabled(), Preferences.FORMATTING_RANGE_ID, Preferences.TEXT_DOCUMENT_RANGE_FORMATTING, null);
+		}
+		if (preferenceManager.getClientPreferences().isCodeLensDynamicRegistrationSupported()) {
+			toggleCapability(preferenceManager.getPreferences().isCodeLensEnabled(), Preferences.CODE_LENS_ID, Preferences.TEXT_DOCUMENT_CODE_LENS, new CodeLensOptions(true));
+		}
+		if (preferenceManager.getClientPreferences().isSignatureHelpDynamicRegistrationSupported()) {
+			toggleCapability(preferenceManager.getPreferences().isSignatureHelpEnabled(), Preferences.SIGNATURE_HELP_ID, Preferences.TEXT_DOCUMENT_SIGNATURE_HELP, SignatureHelpHandler.createOptions());
+		}
+		if (preferenceManager.getClientPreferences().isRenameDynamicRegistrationSupported()) {
+			toggleCapability(preferenceManager.getPreferences().isRenameEnabled(), Preferences.RENAME_ID, Preferences.TEXT_DOCUMENT_RENAME, null);
```

Comment:
```java
if (preferenceManager.getPreferences().isRenameEnabled()) {
	registerCapability(Preferences.RENAME_ID, Preferences.TEXT_DOCUMENT_RENAME);
} else {
	unregisterCapability(Preferences.RENAME_ID, Preferences.TEXT_DOCUMENT_RENAME);
}
``` 
should have been removed back in https://github.com/eclipse/eclipse.jdt.ls/commit/3c872459836f8fb921decb59241f2a30cd84768d#diff-836a9345804342fc811e247515539e98R265 

It's replaced by the call to `toggleCapability `